fl4health/utils/client.py

```diff
@@ -0,0 +1,130 @@
+        return {key: value.to(device) for key, value in data.items()}
+    else:
+        raise TypeError(
+            "data must be of type torch.Tensor or Dict[str, torch.Tensor]. \
+                If definition of TorchInputType or TorchTargetType has \
+                changed this method might need to be updated or split into \
+                two"
+        )
+
+
+def is_empty_batch(input: TorchInputType) -> bool:
+    """
+    Check whether input, which represents a batch of inputs to a model, is empty.
+
+    Args:
+        input (TorchInputType): Input batch. input can be of type torch.Tensor or Dict[str, torch.Tensor], and in the
+        latter case, the batch is considered to be empty if all tensors in the dictionary have length zero.
+
+    Raises:
+        TypeError: Raised if input is not of type torch.Tensor or Dict[str, torch.Tensor].
+        ValueError: Raised if input has type Dict[str, torch.Tensor] and not all tensors within the dictionary have
+            the same size.
+
+    Returns:
+        bool: True if input is an empty batch.
+    """
+    if isinstance(input, torch.Tensor):
+        return len(input) == 0
+    elif isinstance(input, dict):
+        input_iter = iter(input.items())
+        _, first_val = next(input_iter)
+        first_val_len = len(first_val)
+        if not all(len(val) == first_val_len for _, val in input_iter):
+            raise ValueError("Not all tensors in the dictionary have the same size.")
+        else:
+            return first_val_len == 0
+    else:
+        raise TypeError("Input must be of type torch.Tensor or Dict[str, torch.Tensor].")
+
+
+def clone_and_freeze_model(model: nn.Module) -> nn.Module:
+    """
+    Creates a clone of the model with frozen weights to be used in loss calculations so the original model is
+    preserved in its current state.
+
+    Args:
+        model (nn.Module): Model to clone and freeze
+    Returns:
+        nn.Module: Cloned and frozen model
+    """
+
+    cloned_model = copy.deepcopy(model)
+    for param in cloned_model.parameters():
+        param.requires_grad = False
+    cloned_model.eval()
+
+    return cloned_model
+
+
+def maybe_progress_bar(iterable: Iterable, display_progress_bar: bool) -> Iterable:
+    """
+    Used to print progress bars during client training and validation. If self.progress_bar is false, just returns
+    the original input iterable without modifying it.
+
+    Args:
+        iterable (Iterable): The iterable to wrap
+        display_progress_bar (bool): Whether to actually wrap the iterable for progress bar display or not.
+
+    Returns:
+        Iterable: _description_
+    """
+    if not display_progress_bar:
+        return iterable
+    else:
+        # Create a clean looking tqdm instance that matches the flwr logging
+        kwargs: Any = {
+            "leave": True,
+            "ascii": " >=",
+            # "desc": f"{LOG_COLORS['INFO']}INFO{LOG_COLORS['RESET']} ",
```

Comment:
Is this supposed to be here? If so, maybe some context on why its commented out but we are still keeping it

Reply:
hmmm...good catch. I'm not sure why this was commented out. It's also commented out in the original function being migrated here. I'll drop it for now. We have it in the git history if we need to retrieve it.